src/render/queue.rs

```diff
@@ -0,0 +1,63 @@
+use core::sync::atomic::{AtomicU32, Ordering};
+
+use cortex_m::asm::wfe;
+
+/// A simple SPSC queue of u32 values.
+///
+/// The implementation is specialized to the use case of scheduling line
+/// renders on the same core as the interrupt.
+pub struct Queue<const SIZE: usize> {
+    rd_ix: AtomicU32,
+    wr_ix: AtomicU32,
+    buf: [AtomicU32; SIZE],
+}
+
+impl<const SIZE: usize> Queue<SIZE> {
+    pub const fn new() -> Self {
+        const ZERO: AtomicU32 = AtomicU32::new(0);
+        Queue {
+            rd_ix: ZERO,
+            wr_ix: ZERO,
+            buf: [ZERO; SIZE],
+        }
+    }
+
+    /// Push a value, assuming queue is not full.
+    pub fn push_unchecked(&self, val: u32) {
+        let wr_ix = self.wr_ix.load(Ordering::Relaxed);
+        let next = (wr_ix + 1) % SIZE as u32;
+        self.buf[wr_ix as usize].store(val, Ordering::Relaxed);
+        self.wr_ix.store(next, Ordering::Release);
+    }
```

Comment:
Looking at the usage of `Queue` a bit more, I see that it is used in a static, so using `Atomic`s gives interior mutability while keeping the struct `Sync`. But this may not be what we really want. For us it would be nice to have something like `thread_local`s where we can have a static that can be `!Sync` for just one core.

We could make an abstraction like that (albeit at runtime using rp2040's [CPUID register](https://docs.rs/rp2040-pac/latest/rp2040_pac/ppb/cpuid/struct.CPUID_SPEC.html) to either: ensure access to the queue is only done through one core, or create one queue per core and select the queue depending on the current core. But as this queue is only used on one of the cores, it would not make sense to have a second queue for core 0.

Reply:
I believe this code is correct as written. It's intended to be a single-producer, single-consumer queue, so doing the math unatomically is not racy. That said, that's not enforced in the Rust type system (or dynamically). The usual way to do that is to have a separate `Sender` and [`Receiver`](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html) type, and make then not-clone and `!Sync` (but still `Send`). That works when the channel is allocated dynamically, but I'm not sure if it can be expressed for something statically allocated.

I've personally been prioritizing performance over soundness, so have been choosing not to allocate (or do a runtime check) if that's a choice. The DMA magic under the hood is fundamentally extremely unsafe, and it's very difficult to model complex lifetimes in Rust - in particular, resources like palettes and fonts need to be valid until the display list that references them is no longer being scanned out.

I'm open to making things safer, but not enthusiastic about adding complexity or runtime overhead. That's a stylistic choice, though, and some potential users may value not having to worry about safety on the application side.

Reply:
That's fully understandable, I think for now this is a good abstraction, and if later we find a better way to model the system in a sound manner, then this will be worth revising.